src/clp_notification_monitor/compression_buffer/compression_buffer.py

```diff
@@ -11,75 +23,119 @@ def __init__(
-    ):
+        group_by_paths_under_prefix: Path,
+    ) -> None:
         self._logger: Logger = logger
 
-        self.__path_list: List[PurePath] = []
-        self.__total_buffer_size: int = 0
-        self.__first_path_timestamp: Optional[datetime] = None
-
         self.__max_buffer_size: int = max_buffer_size
         self.__min_refresh_period: timedelta = timedelta(milliseconds=min_refresh_period)
 
+        self.__named_buffers: Dict[str, SingleBuffer] = {}
+        self.__ready_buffers: List[str] = []
+        self.__group_by_paths_under_prefix: Path = group_by_paths_under_prefix
+
+        # TODO: change the granularity of lock to each buffer
         self.__lock: Lock = Lock()
         self.__populate_buffer_cv: Condition = Condition(self.__lock)
 
-    def clear_buffer(self) -> None:
-        self.__path_list = []
-        self.__total_buffer_size = 0
-        self.__first_path_timestamp = None
-
     def append(
         self,
         s3_path: PurePath,
         object_size: int,
         process_timestamp: datetime,
     ) -> None:
         with self.__populate_buffer_cv:
-            self.__total_buffer_size += object_size
-            if self.__first_path_timestamp is None:
-                self.__first_path_timestamp = process_timestamp
-            self.__path_list.append(s3_path)
+            relative_path: PurePath
+            index_name: str
+            try:
+                relative_path = s3_path.relative_to(self.__group_by_paths_under_prefix)
+                index_name = relative_path.parts[0]
+            except (KeyError, ValueError):
+                # This file does not belong to any named buffer
+                index_name = ""
+
+            # Get the correct buffer
+            target_buffer: SingleBuffer
+            try:
+                target_buffer = self.__named_buffers[index_name]
+            except KeyError:
+                target_buffer = SingleBuffer()
+                self.__named_buffers[index_name] = target_buffer
+
+            # Add path to buffer
+            target_buffer.total_buffer_size += object_size
+            if target_buffer.first_path_timestamp is None:
+                target_buffer.first_path_timestamp = process_timestamp
+            target_buffer.path_list.append(s3_path)
+
+            # Notify the thread waiting for a compression job
             self.__populate_buffer_cv.notify_all()
 
     def wait_for_compression_jobs(self) -> None:
         """
         Block the current process until the buffer is populated.
         """
         with self.__populate_buffer_cv:
-            while None is self.__first_path_timestamp:
+            while True:
+                all_buffers_empty: bool = True
+                for buf in self.__named_buffers.values():
+                    if buf.first_path_timestamp is not None:
+                        all_buffers_empty = False
+                        break
+                if not all_buffers_empty:
+                    break
```

Comment:
In general, this is very inefficient. Maybe we should ass a counter to track the number of files pending in the buffers.
At least we should add a `TODO`